### PR TITLE
WIP: Arch aware node info

### PR DESCRIPTION
**Pull request by aleskandro**
/hold
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go [hunk 1]
```diff
@@ -451,27 +454,45 @@ func createTestConfigs(specs ...testSpec) []*testConfig {
...
+				}
+				if err := unstructured.SetNestedStringMap(v.Object, map[string]string{
+					"architecture":    spec.nodeInfo.Architecture,
+					"operatingSystem": spec.nodeInfo.OperatingSystem,
+				}, "status", "nodeInfo"); err != nil {
+					panic(err)
+				}
+			}
+		}
+
```
Comment (aleskandro) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go]: Hi @elmiko, should we add that field only to machineTemplates or also to machineSets and machineDeployments? I don't see status.capacity set for machineSets and machineDeployments, but I also didn't find tests using a non-nil value for capacity.

Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go]: no, the status.capacity should only come from the infrastructure machine templates.

for machinesets and machinedeployments, the annotations would be used.

Comment (aleskandro) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go]: do we have tests in place that use a non-nil capacity and machine templates?

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go [hunk 2]
```diff
@@ -465,13 +466,34 @@ func createTestConfigs(specs ...testSpec) []*testConfig {
...
 		} else {
 			klog.V(4).Infof("not adding capacity")
 		}
 
+		if spec.nodeInfo != nil {
+			for _, v := range []*unstructured.Unstructured{
+				config.machineSet,
+				config.machineDeployment,
+				config.machineTemplate,
+			} {
```
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go]: similar to the previous block where we set the capacity, the nodeInfo only needs to be present on the `machineTemplate`. it does not need to be on the `machineSet` or `machineDeployment`.

essentially, the mechanics of the test work by getting the machine template resource referenced by a machineset or machinedeployment. then it uses the node info from the machine template to populate the information.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go [hunk 3]
```diff
@@ -304,6 +304,17 @@ func (r unstructuredScalableResource) InstanceCapacity() (map[corev1.ResourceNam
 	return capacity, nil
 }
 
+// InstanceSystemInfo sets the nodeSystemInfo from the infrastructure reference resource.
+// If the infrastructure reference resource is not found, returns nil.
+func (r unstructuredScalableResource) InstanceSystemInfo() *apiv1.NodeSystemInfo {
+	infraObj, err := r.readInfrastructureReferenceResource()
```
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:330]: something i am concerned about is the notion that we will be calling this function twice for every node template we create. each call to this function results in another call to the api server to get the infrastructure machine template. i think we need to find a way to combine the call required in this function with the call we make in the `InstanceCapacity` call.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go [hunk 4]
```diff
@@ -543,3 +552,83 @@ func TestCanScaleFromZero(t *testing.T) {
...
+			"",
+		},
+		{
+			"with operating system in machine template's status' nodeInfo, the system info is filled in the scalable resource",
+			map[string]string{
+				operatingSystemStatusKey: linux,
+			},
+			"",
+			linux,
+		},
```
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go:637]: we should have some tests for the negative assertion cases where the information is not present.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go (hunk 2 above)
Comment (aleskandro) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go]: Ops. Thanks, I forgot I did not revert this.

Comment (aleskandro) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go]: Fixed.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go (hunk 4 above)
Comment (aleskandro) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go:637]: Done

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go (hunk 3 above)
Comment (aleskandro) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:330]: To keep having two separate functions and avoid coupling, I've made a change to the readInfrastructureReferenceResource to cache the infraObj via lazy loading. Wdyt?

**Comment by elmiko** [comment]
apologies @aleskandro , i have been caught on some bug triaging and have not been able to review this again. i am putting it on my queue for next week. hopefully i can review on tuesday.

**Comment by aleskandro** [comment]
No problem @elmiko. I rebased and tests are now passing.

PTAL when you get the chance.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go [hunk 5]
```diff
@@ -465,13 +465,25 @@ func createTestConfigs(specs ...testSpec) []*testConfig {
...
+			klog.V(4).Infof("adding capacity to machine template")
+
 			if err := unstructured.SetNestedStringMap(config.machineTemplate.Object, spec.capacity, "status", "capacity"); err != nil {
 				panic(err)
 			}
 		} else {
 			klog.V(4).Infof("not adding capacity")
 		}
 
+		if spec.nodeInfo != nil {
```
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller_test.go]: since we have the log for adding capacity, let's add a similar log here for adding node info.

```
klog.V(4).Infof("adding nodeInfo to machine template")
```

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go [hunk 6]
```diff
@@ -392,6 +406,17 @@ func (r unstructuredScalableResource) InstanceDRADriver() string {
 }
 
 func (r unstructuredScalableResource) readInfrastructureReferenceResource() (*unstructured.Unstructured, error) {
+	// Cache w/ lazy loading of the infrastructure reference resource.
+	r.infraMutex.RLock()
+	if r.infraObj != nil {
```
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:409]: i like the idea here, but i'm not sure if this will cause undue issues. one thing a user can do is change the infra object reference in a scalable resource (for example to trigger a new rollout). if i'm reading this correctly, it would be possible for a user to change the instance type by using a new infra object and then it would not get picked up in this record (unless the autoscaler restarted). do we need an age factor for this cache perhaps?

**Review by elmiko — COMMENTED** [review]
this mostly makes sense to me, i have a couple questions.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go (hunk 6 above)
Comment (aleskandro) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:409]: Hi @elmiko, in the proposal here, I did not consider the scalableResource object as 'cached'. Is this the case or is the scalableResource instantiated ad-hoc (with the most up to date version) when the reconciliation loop runs?

Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:409]: i need to doublecheck the mechanism, but it should be reloaded when needed. i'm not sure if it's possible for the object to exist for a long time though.

Comment (aleskandro) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:409]: Tracing back the creation, I lead to:

1. machineController.nodeGroups() https://github.com/kubernetes/autoscaler/blob/f4c2fdebeb092bfea4aa709c0624296085c12b47/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go#L754-L774 (it makes use of informer's listers) 
1. machineController.nodeGroupForNode (still relying on informers) https://github.com/kubernetes/autoscaler/blob/f4c2fdebeb092bfea4aa709c0624296085c12b47/cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go#L775-L798

it seems ok to me to store the cache as a field of the scalableResource struct as the cache would not live beyond the lifecycle of a specific instance requested by the MachineController.

Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:409]: perfect, thank you. that satisfies my curiosity.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go [hunk 7]
```diff
@@ -572,3 +581,89 @@ func TestCanScaleFromZero(t *testing.T) {
...
```
Comment (jackfrancis) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go:637]: let's add one more test scenario where both arch _and_ os info is present in the NodeInfo object (and validate that both are present in the resource that it returns

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go (hunk 6 above)
Comment (aleskandro) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:409]: Thanks for the review and support @elmiko :)

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go (hunk 7 above)
Comment (aleskandro) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured_test.go:637]: Hi @jackfrancis, thanks for the review. Done!

**Comment by elmiko** [comment]
thanks for the updates @aleskandro , i think this is looking good.

/lgtm

i'd like to give @jackfrancis a chance to review again, but i have a feeling we can approve this soon.

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go [hunk 8]
```diff
@@ -380,8 +385,19 @@ func (ng *nodegroup) TemplateNodeInfo() (*framework.NodeInfo, error) {
...
+		nsiLabels[corev1.LabelArchStable] = nsi.Architecture
+		nsiLabels[corev1.LabelOSStable] = nsi.OperatingSystem
+	}
+
+	// The order of priority is:
+	// - Labels set in existing nodes for not-autoscale-from-zero cases
+	// - Labels set in the labels capacity annotation of machine template, machine set, and machine deployment.
+	// - Values in the status.nodeSystemInfo of MachineTemplates
+	// - Generic/default labels set in the environment of the cluster autoscaler
+	labels := cloudprovider.JoinStringMaps(buildGenericLabels(nodeName), nsiLabels, ng.scalableResource.Labels())
```
Comment (jackfrancis) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go:400]: This is a bit academic, but now that there's a chance we'll be passing in a nil `map[string]string` to `cloudprovider.JoinStringMaps`, let's update the existing `TestJoinStringMaps` UT to validate that use case. 

Something like

```sh
$ git diff
diff --git a/cluster-autoscaler/cloudprovider/util_test.go b/cluster-autoscaler/cloudprovider/util_test.go
index ba23d6180..ee99f2bcb 100644
--- a/cluster-autoscaler/cloudprovider/util_test.go
+++ b/cluster-autoscaler/cloudprovider/util_test.go
@@ -44,9 +44,11 @@ func TestBuildKubeProxy(t *testing.T) {
 }
 
 func TestJoinStringMaps(t *testing.T) {
+       emptyMapBeginning := make(map[string]string)
        map1 := map[string]string{"1": "a", "2": "b"}
        map2 := map[string]string{"3": "c", "2": "d"}
        map3 := map[string]string{"5": "e"}
-       result := JoinStringMaps(map1, map2, map3)
+       emptyMapEnd := make(map[string]string)
+       result := JoinStringMaps(emptyMapBeginning, map1, map2, map3, emptyMapEnd)
        assert.Equal(t, map[string]string{"1": "a", "2": "d", "3": "c", "5": "e"}, result)
 }
```

Comment (aleskandro) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_nodegroup.go:400]: Hi @jackfrancis , thanks for the review and apologies for the long delay here. Just committed.

**Review by elmiko — COMMENTED** [review]
/lgtm

@jackfrancis are you ok to approve?
